### ISSUE TITLE
HTTP: post data doesn't get sent over.

### ISSUE THREAD

**Issue by jelmersnoeck** [github/hubot#934]
When trying to do a post request to an endpoint, the post data doesn't get sent over.

```coffee
data = { foo: 'bar' }
postData = JSON.stringify(data)

msg.http('http://127.0.0.1:3000/apiendpoint')
  .headers('Content-Type': 'application/json', 'Content-Length': postData.length)
  .post(postData) (err, res, body) ->
    msg.send(body)
```

I have my localhost set up to raise an exception with the post data, which results in `{}`

In case of nested attributes, the following happened:

```coffee
postData = 
  foo:
    bar: 'value'

# response
{ 'foo': '' }
```

I've tried several other methods as well:

- `querystring.stringify` instead of `JSON.stringify`
- `.query(postData)` instead of `.post(postData)`. Here it actually sends the key over but not the value
- without headers
- nested data (this is actually what I want in my script)

**Comment by technicalpickles** [comment]
That is pretty weird. Looking at the code snippet you have here, I can't see anything obviously wrong. One thing I would suggest is checking for `err`, like:

```coffeescript
  .post(postData) (err, res, body) ->
    if err?
      return robot.emit 'error', err, msg

    msg.send(body)
```

The only thing I can think of is that there was an error (like not being able to connect). When an error happens, the body is going to be empty.

**Comment by jelmersnoeck** [comment]
I'm sorry, I forgot to update this.

I managed to get this working a bit back with the Node HTTPS module. This had issues at first as well, using a nested hash (I have updated the issue to reflect this). Instead of using `bar: 'value'` I used `'bar': 'value'` which did the trick.

I just retried with the Node scoped HTTP client and this seems to work as well (by just using `bar: 'value'`) which makes me suspect - as you suggested - that this might have been a connection error.

I will look into this again to see if I can reproduce and check on errors if there are any.
